ls1intum/Artemis
# athene load balancer

### Checklist

### Motivation and Context
Artemis currently calls each Athene-component (segmentation, embedding, clustering) sequentially by itself. This blocks an Artemis thread as long as these calculations are running and does not support load-balancing of the calculation. Since especially the embedding-process is a long-running job, we introduced a load-balancing-component to the Athene system, which will handle a request for automatic feedback suggestion calculation and will send the results back to Artemis via a callbackUrl (${server.url}/api/athene-result({exerciseId}).

### Description
The changes in this PR support the conversion of the existing stand-alone Athene components to the new load-balanced system. Thus, Artemis will only send one single request to the Athene system to invoke the calculation of automatic feedback suggestions for text exercises. This PR is related to [PR #27 in Athene](https://github.com/ls1intum/Athene/pull/27).

### Steps for Testing
1. Start the Athene-system in the state of [PR #27](https://github.com/ls1intum/Athene/pull/27) using `docker-compose up -d`
2. If needed, adjust the ports used by Athene to unused ones on your machine using the `.env`-file:

- TRAEFIK_DASHBOARD_PORT=8081 (unimportant for this PR, this gives status information)
- TRAEFIK_HTTP_PORT=80
- TRAEFIK_HTTPS_PORT=443 (unimportant for this PR, since we use unencrypted HTTP in non-production setups)

3. Adjust the variable `server.url` in the Artemis config (e.g. in `application-local.yml`): Set this variable to the IP of your machine in its current local network (e.g. 192.168.1.20, **NOT localhost**). Otherwise, the Athene system will not be able to reach out to Artemis out of its docker bridge-network, if all is running on the same machine. You can get this IP using `ipconfig` (Windows), `ifconfig` (Linux and macOS), or by inspecting your network adapter elsewhere in the OS.
4. If you changed the `TRAEFIK_HTTP_PORT` of Athene in step 2, make sure to adjust the variable `artemis.athene.submit-url` corresponding to your change
5. Start Artemis with the profiles `dev,bamboo,bitbucket,jira,artemis,scheduling,local,athene`
6. Log in to Artemis
7. Navigate to Course Management
8. Select a course, create a new text exercise and enable "Automatic assessment suggestions enabled"
9. Create at least 10 submissions for this exercise before the deadline
10. Let the deadline pass
11. Inspect the logs and check for errors of the Athene components using

- `docker logs -f athene-load-balancer` (Interaction with Artemis and all calculation components)
- `docker logs -f athene-segmentation` (1st calculation step)
- `docker logs -f athene-embedding` (2nd calculation step)
- `docker logs -f athene-clustering` (3rd calculation step)

12. Check the database tables `text_block` and `text_cluster` for entries
13. Check the automatic feedback suggestions by starting an assessment as a tutor of the course the created exercise of step 8 belongs to

### Test Coverage
* AtheneService.java: 92%
* TextAssessmentConflictService.java: 100%
